File: 1570-final-prices-with-a-special-discount-in-a-shop/1570-final-prices-with-a-special-discount-in-a-shop.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        stack<int> st;
        for (int i = prices.size() - 1; i >= 0; i--) {
            while (!st.empty() && st.top() > prices[i]) {
                st.pop();
            }
            int ele = prices[i];
            if (!st.empty()) prices[i] = prices[i]-st.top();
            st.push(ele);
        }
        return prices;
    }
};

File: 1570-final-prices-with-a-special-discount-in-a-shop/1570-final-prices-with-a-special-discount-in-a-shop.java
class Solution {
    public int[] finalPrices(int[] prices) {
        int n = prices.length;
        int[] discPrices = new int[n];
        Stack<Integer> st = new Stack<>();
        for(int i = n-1; i>=0; i--){
            while(!st.isEmpty() && prices[i]<st.peek()) st.pop();
            if(st.isEmpty()) discPrices[i] = prices[i];
            else discPrices[i] = prices[i] - st.peek();
            st.push(prices[i]);
        }
        return discPrices;
    }
}